File: apps/desktop/src/lib/creds.ts
/**
 * Feed credentials — the frontend half of `src-tauri/src/creds.rs`.
 *
 * Three calls, each a thin `invoke` over the Rust command of the same name,
 * so the secret crosses exactly one boundary (webview → Rust → keychain) and
 * is never written by JavaScript to anything that persists. `localStorage`
 * holds one thing only: the **name** of the last vendor used, which is not a
 * secret and is what lets the shell know which keychain entry to open on the
 * next launch.
 *
 * ## Outside Tauri
 *
 * `pnpm dev` runs this UI as a plain page with no Rust behind it. There the
 * store is a `Map` that lives until reload and `PERSISTENT` is `false`, which
 * the form shows. The alternative — `localStorage` as a dev fallback — is a
 * plaintext credential on disk on a developer machine, and the W3D1 rule does
 * not have a dev-mode exception.
 *
 * ## Shape
 *
 * `FeedCreds` is `{ vendor; [field]: string | undefined }` and stays that way
 * until the Ironbeam account (`plans/current.md` #9) exists and its fields
 * are known. This module and the form take the map as-is; the day the fields
 * are known, the form gains labels and this file does not change.
 */

import { invoke, isTauri } from "@tauri-apps/api/core";

import type { FeedCreds } from "./engine/types";

/** Whether saved credentials survive a restart on this build. */
export const PERSISTENT: boolean = isTauri();

const LAST_VENDOR_KEY = "ti.feed.vendor";

const memory = new Map<string, string>();

/** Everything but `vendor`, as the JSON the keychain entry holds. */
function fieldsOf(creds: FeedCreds): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [k, v] of Object.entries(creds)) {
    if (k !== "vendor" && typeof v === "string" && v !== "") out[k] = v;
  }
  return out;
}

export async function saveFeedCreds(creds: FeedCreds): Promise<void> {
  const json = JSON.stringify(fieldsOf(creds));
  if (PERSISTENT) {
    await invoke("creds_save", { vendor: creds.vendor, fieldsJson: json });
  } else {
    memory.set(creds.vendor, json);
  }
  localStorage.setItem(LAST_VENDOR_KEY, creds.vendor);
}

export async function loadFeedCreds(vendor: string): Promise<FeedCreds | null> {
  const json = PERSISTENT
    ? await invoke<string | null>("creds_load", { vendor })
    : (memory.get(vendor) ?? null);
  if (json === null) return null;
  return { vendor, ...(JSON.parse(json) as Record<string, string>) };
}

export async function clearFeedCreds(vendor: string): Promise<void> {
  if (PERSISTENT) {
    await invoke("creds_clear", { vendor });
  } else {
    memory.delete(vendor);
  }
  if (localStorage.getItem(LAST_VENDOR_KEY) === vendor) {
    localStorage.removeItem(LAST_VENDOR_KEY);
  }
}

/** The vendor whose credentials were last saved, or null. Not a secret. */
export function lastVendor(): string | null {
  return localStorage.getItem(LAST_VENDOR_KEY);
}

File: apps/desktop/src-tauri/src/creds.rs
//! Feed credentials, in the OS keychain and nowhere else.
//!
//! W3D1 (`plans/team/prathamesh/README.md`): *"Credentials to the OS
//! keychain; never plaintext on disk, never to our servers."* This module is
//! the whole of that sentence on the Rust side. Three commands, one keychain
//! entry per vendor, the fields serialised as JSON in the entry's secret.
//!
//! ## Why one entry per vendor and not one per field
//!
//! `keyring` can set, get and delete an entry by `(service, account)`; it
//! cannot list entries. A field-per-entry layout would need a separate index
//! of which fields exist -- and an index of credential field names on disk is
//! plaintext about the credentials, which is the thing this exists to avoid.
//! One JSON blob under the vendor's name needs no index.
//!
//! ## What this module does not know
//!
//! Which fields a vendor has. `FeedCreds` in `engine/types.ts` is
//! deliberately `{ vendor; [field]: string }` until the Ironbeam account
//! (`plans/current.md` #9) exists and its fields are known, and this module
//! stores whatever map it is given. The shape is the UI's to firm up later;
//! storage does not care.
//!
//! ## Native backends only
//!
//! `Cargo.toml` enables `apple-native`, `windows-native` and
//! `sync-secret-service`, and none of the crate's file-based fallbacks. If
//! the platform has no keychain, `Entry::new` or `set_password` fails and the
//! frontend shows that rather than quietly writing a file.

use keyring::Entry;

/// The keychain "service". Every entry this app writes is under this name,
/// so a trader looking in Keychain Access sees exactly what we hold.
// Renamed from `trading-intelligence.feed` with the Vision Hub rebrand. An
// entry saved under the old name is not migrated: the form asks for the
// credentials again, which is one paste, and a migration that copies a
// secret between keychain entries is code that handles a secret for no
// gain.
const SERVICE: &str = "com.visionhub.desktop.feed";

fn entry(vendor: &str) -> Result<Entry, String> {
    if vendor.trim().is_empty() {
        return Err("vendor must not be empty".into());
    }
    Entry::new(SERVICE, vendor).map_err(|e| format!("keychain unavailable: {e}"))
}

/// Store `fields_json` as the secret for `vendor`, replacing any previous one.
#[tauri::command]
pub fn creds_save(vendor: &str, fields_json: &str) -> Result<(), String> {
    entry(vendor)?
        .set_password(fields_json)
        .map_err(|e| format!("keychain write failed: {e}"))
}

/// The stored fields for `vendor`, or `None` if nothing is stored.
#[tauri::command]
pub fn creds_load(vendor: &str) -> Result<Option<String>, String> {
    match entry(vendor)?.get_password() {
        Ok(json) => Ok(Some(json)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(format!("keychain read failed: {e}")),
    }
}

/// Remove the stored fields for `vendor`. Removing nothing is not an error.
#[tauri::command]
pub fn creds_clear(vendor: &str) -> Result<(), String> {
    match entry(vendor)?.delete_credential() {
        Ok(()) | Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(format!("keychain delete failed: {e}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Round-trips through the real keychain of whatever machine runs the
    /// suite. The item is created and deleted by the same binary, which on
    /// macOS is the case that does not prompt.
    #[test]
    fn save_load_clear_round_trip() {
        let vendor = format!("test-{}", std::process::id());
        assert_eq!(creds_load(&vendor).unwrap(), None);

        creds_save(&vendor, r#"{"apiKey":"k","secret":"s"}"#).unwrap();
        assert_eq!(
            creds_load(&vendor).unwrap().as_deref(),
            Some(r#"{"apiKey":"k","secret":"s"}"#)
        );

        creds_save(&vendor, r#"{"apiKey":"k2"}"#).unwrap();
        assert_eq!(creds_load(&vendor).unwrap().as_deref(), Some(r#"{"apiKey":"k2"}"#));

        creds_clear(&vendor).unwrap();
        assert_eq!(creds_load(&vendor).unwrap(), None);
        // Clearing twice is fine.
        creds_clear(&vendor).unwrap();
    }

    #[test]
    fn empty_vendor_is_refused() {
        assert!(creds_save("  ", "{}").is_err());
        assert!(creds_load("").is_err());
    }
}
